akd_audit_quorum/src/crypto/mod.rs

```diff
@@ -0,0 +1,309 @@
+    fn clone(&self) -> Self {
+        Self {
+            components: self.components,
+        }
+    }
+}
+
+impl QuorumKeyShard {
+    pub(crate) fn build_from_vec_vec_vec(
+        data: Vec<Shard>,
+    ) -> Result<Vec<Self>, QuorumOperationError> {
+        let mut results = vec![];
+
+        for shards in data.into_iter() {
+            let mut formatted_shards: Vec<[u8; SHARE_SIZE]> = vec![];
+            for shard in shards.into_iter() {
+                formatted_shards.push(shard.try_into().map_err(|_| {
+                    QuorumOperationError::Sharding(format!(
+                        "Unable to convert shard vec into array of len {}",
+                        DATA_SIZE
+                    ))
+                })?)
+            }
+            let formatted_shard = Self {
+                components: formatted_shards.try_into().map_err(|_| QuorumOperationError::Sharding(format!("Unable to format vector of shards into quorum key shard struct with {} components", QUORUM_KEY_NUM_PARTS)))?
+            };
+            results.push(formatted_shard);
+        }
+
+        Ok(results)
+    }
+
+    /// Flatten the crypto shard into a single array of the components in-order
+    pub fn flatten(&self) -> [u8; SHARE_SIZE * QUORUM_KEY_NUM_PARTS] {
+        let mut data = [0u8; SHARE_SIZE * QUORUM_KEY_NUM_PARTS];
+        #[allow(clippy::needless_range_loop)]
+        for part_i in 0..QUORUM_KEY_NUM_PARTS {
+            let start = part_i * SHARE_SIZE;
+            let end = (part_i + 1) * SHARE_SIZE;
+            data[start..end].clone_from_slice(&self.components[part_i]);
+        }
+        data
+    }
+
+    /// Inflate a decrypted crypto shard from the flattened components
+    pub fn inflate(
+        raw: [u8; SHARE_SIZE * QUORUM_KEY_NUM_PARTS],
+    ) -> Result<Self, QuorumOperationError> {
+        let mut components = [[0u8; SHARE_SIZE]; QUORUM_KEY_NUM_PARTS];
+        #[allow(clippy::needless_range_loop)]
+        for part_i in 0..QUORUM_KEY_NUM_PARTS {
+            let start = part_i * SHARE_SIZE;
+            let end = (part_i + 1) * SHARE_SIZE;
+
+            let slice = raw[start..end].to_vec();
+            components[part_i] = slice.try_into().map_err(|_| {
+                QuorumOperationError::Sharding(
+                    "Unable to deserialize raw binary vec to QuorumKeyShard".to_string(),
+                )
+            })?;
+        }
+
+        Ok(Self { components })
+    }
+}
+
+/// Represents an encrypted quorum key shard which is encrypted with
+/// the public key of the reliant party. These shards cannot be decoded in
+/// user-space, and must be handled WITHIN the crypto layer such that they
+/// public key cannot be reconstructed by "non secure" memory. The private key
+/// which can decrpyt the shards is the node's "transient" key used for communication
+/// channels, and never is exposed outside of the cryptographer service
+#[derive(Clone)]
+pub struct EncryptedQuorumKeyShard {
```

Comment:
An alternative to this sharding approach would be to use the frost-dalek library: https://docs.rs/frost-dalek/0.2.3/frost_dalek/

The advantage being that no single quorum participant needs to be responsible for fully constructing the signing key. Instead, each quorum participant uses their key shard to produce a "signature shard", and then anyone can aggregate the signature shards together to compose the actual signature.

Thoughts on integrating frost-dalek into here?

Reply:
We're in the process of resurrecting this avenue. We'll definitely look into this usage! We just need to assess growth & shrinkage of the quorum over time, but if there's a solution there it's an attractive option! Thanks for the pointer :)